### ISSUE TITLE
Documentation: generate and publish multiple versions of the documentation.

### ISSUE THREAD

**Issue by klihub** [intel/cri-resource-manager#418]
Change the publishing github workflow to
  - discover each tagged release (with sphinx docs bits)
  - generate and publish documentation for each tagged release
  - generate and publish documentation for the rolling main branch
  - record (and publish) the exact version/SHA1 for each version of the documentation
  - nuke sources from the gh-pages branch

**Comment by klihub** [comment]
See #405 for related discussion and background.

**Comment by marquiz** [comment]
I was thinking about doing it so that have a separate folder for every release branch, like `0.4`, `0.5` etc plus one for master. Whenever a tag is pushed build the documentation and update the corresponding release folder. Similarly update the master folder whenever master branch is updated.

In this scheme we would not always re-build all documentation versions but only when something is updated. WDYT?

**Comment by klihub** [comment]
> I was thinking about doing it so that have a separate folder for every release branch, like `0.4`, `0.5` etc plus one for master. Whenever a tag is pushed build the documentation and update the corresponding release folder. Similarly update the master folder whenever master branch is updated.

Yes, I also think something like that is needed. But I think we need to generate master-like rolling type documentation for
the HEAD of each release branch if it's SHA1 is not same as one of the tags. IOW we should be prepared to generate
documentation for tags (locked releases), for rolling release branches' HEADs, and for rolling master HEAD. We should
assume that tags are locked, IOW they do not change, but verify this (hence recording the SHA1 for tags is necessary) and regenerate the corresponding documentation if we bit the bullet and re-tagged after a screw up.
 
> In this scheme we would not always re-build all documentation versions but only when something is updated. WDYT?

Yes, I also think that we need to generate docs to separate directories and only regenerate documentation for a tag,
release branch, or master if it is necessary due to changes.

**Comment by marquiz** [comment]
> But I think we need to generate master-like rolling type documentation for
> the HEAD of each release branch if it's SHA1 is not same as one of the tags. IOW we should be prepared to generate
> documentation for tags (locked releases), for rolling release branches' HEADs, and for rolling master HEAD.

Hmm, I don't necessarily see a point in generating documentation for the "rolling release branches". I think that anything ending up there (I mean any PRs getting merged into the release branches) should be released i.e. tagged very fast. If we want rolling releases then probably we should really make them behave like that, e.g. with automatic tagging. Let's say do it periodically, e.g. every night automatically create a new tag for any unreleased content on release branches.

My thinking was to only maintain documentation separately for `master` and the release major versions. 
```
 gh-pages
    ├── master
    ├── v0.4
    ├── v0.5
    ...
```
E.g. `v0.4.x` documentation would be overwritten when `v0.4.(x+1)` get's tagged




> We should
> assume that tags are locked, IOW they do not change, but verify this (hence recording the SHA1 for tags is necessary) and regenerate the corresponding documentation if we bit the bullet and re-tagged after a screw up.

We really must not force push over existing tags! Especially if the user-base grows. That is really a major  cause of confusion. I really think that if something needs to be fixed after the fact then just swallow one's pride, accept your mistake and make a quick point release. With automatic tagging that would be almost a no-op. I'd like to keep the doc generation script as simple as reasonably possible: act on the git tag/hash that the job received and nothing else.

Thus, I would not script any automatic checking or regeneration of the whole documentation. I feel this would complicate the doc generation script and it's maintenance without much real world benefit. Moreover, with automatic doc generation we would re-generate docs even on force pushing over existing tags so that should be covered

**Comment by klihub** [comment]
> > But I think we need to generate master-like rolling type documentation for
> > the HEAD of each release branch if it's SHA1 is not same as one of the tags. IOW we should be prepared to generate
> > documentation for tags (locked releases), for rolling release branches' HEADs, and for rolling master HEAD.
> 
> Hmm, I don't necessarily see a point in generating documentation for the "rolling release branches". I think that anything ending up there (I mean any PRs getting merged into the release branches) should be released i.e. tagged very fast. If we want rolling releases then probably we should really make them behave like that, e.g. with automatic tagging. Let's say do it periodically, e.g. every night automatically create a new tag for any unreleased content on release branches (assuming all automatic release checks/criteria pass muster).

> My thinking was to only maintain documentation separately for `master` and the release major versions.

> 
> ```
>  gh-pages
>     ├── master
>     ├── v0.4
>     ├── v0.5
>     ...
> ```
> 
> E.g. `v0.4.x` documentation would be overwritten when `v0.4.(x+1)` get's tagged

TBH, I don't see what that would really buy us in terms of simpler tooling/work saved, while
I do see how it would build in several implicit assumption/limitations:
- limitation: not being able to host documentation for multiple minor-release versions
- assumption: all PR's merged to a release branch are extremely short-lived, with a tag very soon appearing on the path between the merge commit and the tip/HEAD of the branch

And my biggest problem with that approach would be that, I really don't see why should we treat release branches so differently from master, where in reality IMO a release branch is semantically/essentially a filtered version of master where you are just a bit more picky about which PRs/commits you let onto the branch. Without any implicit related assumptions about tagging or how long-lived a branch is without a tag at the tip.

> 
> > We should
> > assume that tags are locked, IOW they do not change, but verify this (hence recording the SHA1 for tags is necessary) and regenerate the corresponding documentation if we bit the bullet and re-tagged after a screw up.
> 
> We really must not force push over existing tags! Especially if the user-base grows. That is really a major cause of confusion. I really think that if something needs to be fixed after the fact then just swallow one's pride, accept your mistake and make a quick point release. With automatic tagging that would be almost a no-op. I'd like to keep the doc generation script as simple as reasonably possible: act on the git tag/hash that the job received and nothing else.
> 
> Thus, I would not script any automatic checking or regeneration of the whole documentation. I feel this would complicate the doc generation script and it's maintenance without much real world benefit. Moreover, with automatic doc generation we would re-generate docs even on force pushing over existing tags so that should be covered

I agree that we should never do it. It's just that I usually rather build such a limitation into our practices and not into our abilities/tooling. That said, I don't have any strong opinion about this and I think it's also fine to assume we'll never do this again but instead immediately roll another tagged release if we screw up. Or to put it another way, I much rather give this one up than the desire/ability to treat release branches like a potentially long-lived commit-filtered rolling master.

**Comment by klihub** [comment]
That said, I'm not too insistent about this. I can surely live with any approach that allows us to publish docs for multiple versions. If it just the latest tagged versions for any 0.x release + rolling master, then that is already an improvement over what we have today.

**Comment by marquiz** [comment]
Maybe we see the release branches a little differently 😄  I really think that the release (stable) branches are different in nature from the master (development) branch.

I see that the dot releases should really be about small bug/security fixes - possibly rarely about backporting some small feature. Essentially there should be no major/breaking  changes in the functionality and user interface. If there were some major changes those should be indicated in the documentation of that release branch like *Added/changed/new in v0.4.3*. We probably even don't have the bandwidth to actively maintain many release branches. So, documentation for `v0.4.(x+1)` should be "compatible" with `v0.4.x`. I think this is a common paradigm/practice among other projects, too.

We could generate documentation for the "rolling release branches", but, I wouldn't create links for them in the github pages. Users should be advised to use tagged releases only.

**Comment by marquiz** [comment]
> That said, I'm not too insistent about this. I can surely live with any approach that allows us to publish docs for multiple versions. If it just the latest tagged versions for any 0.x release + rolling master, then that is already an improvement over what we have today.

Yeah, let's start with something plain and simple and improve on that. I think it would be time to implement something for a change 😉   I try to get my documentation restructuring PR out and then work on this

**Comment by mythi** [comment]
> Yeah, let's start with something plain and simple and improve on that. 

+1 for having docs generated/updated for what is the latest in `release-x.y` and `master`. it easy to do with `actions/checkout` giving steps for various branches. (one thing to note is that if moving to `checkout@v2`, the `$author/$email` start failing because v2 checks out the latest commit only)

**Comment by klihub** [comment]
> +1 for having docs generated/updated for what is the latest in `release-x.y` and `master`. it easy to do with `actions/checkout` giving steps for various branches. (one thing to note is that if moving to `checkout@v2`, the `$author/$email` start failing because v2 checks out the latest commit only)

But IIUC/according to the documentation this default behavior can be overridden to a full-depth checkout:
```
Only a single commit is fetched by default, for the ref/SHA that triggered the workflow. Set fetch-depth: 0 to fetch all history for all branches and tags.
```

**Comment by klihub** [comment]
#433

**Comment by klihub** [comment]
Implemented and merged by #440 et al.
